packages/tools/src/tools/annotation/planarFreehandROITool/interpolation/interpolateSegmentPoints.ts

```diff
@@ -0,0 +1,59 @@
+import { Types } from '@cornerstonejs/core';
+import { interpolatePoints } from './algorithms/bspline';
+
+/**
+ * Interpolates a segment of points from iniIndex until endIndex.
+ * It uses the knotSampleSize param to calculate which point from points will be used as knot or not.
+ * Lower knotSampleSize value is more points will be used, i.e the interpolation will be harder.
+ * By default it uses b-spline algorithm.
+ * The result total of points is equal to original points.
+ */
+export default function interpolateSegmentPoints(
+  points: Types.Point2[],
+  iniIndex: number,
+  endIndex: number,
+  knotSampleSize: number
+): Types.Point2[] {
+  const segmentSize = endIndex - iniIndex + 1;
+  if (isNaN(segmentSize) || !segmentSize || !knotSampleSize) {
+    return points;
+  }
+
+  const knotsDistance = Math.floor(segmentSize / knotSampleSize);
+
+  const interpolationIniIndex = Math.max(0, iniIndex - 4 * knotsDistance);
+  const interpolationEndIndex = Math.min(
+    points.length - 1,
+    endIndex + 4 * knotsDistance
+  );
+
+  const segmentPointsUnchangedBeg = points.slice(0, interpolationIniIndex);
+
+  const segmentPointsUnchangedEnd = points.slice(
+    interpolationEndIndex + 1,
+    points.length
+  );
+
+  const knotsIndexes = [];
+  let intIt = 0;
+  let nextKnotIndex = intIt * knotsDistance + interpolationIniIndex;
+
+  while (nextKnotIndex <= interpolationEndIndex) {
+    knotsIndexes[intIt] = nextKnotIndex;
+    intIt++;
+    nextKnotIndex = intIt * knotsDistance + interpolationIniIndex;
```

Comment:
This calculation ends up putting all the fractional values into the last element, which produces an irregularity in the bspline result.  You should use the floor value here with the fractional knotsDistance.
Let me give an example - suppose the knotsDistance is 3 and the total length is 10 for the distance - you get:
0,3,6,9,10
as your points - note how 9 and 10 only differ by 1.  If you do this fractionally, you get:
0,2,5,8,10  so the split is on the first and last element.      It also means you can run this as a for loop, since you can calculate exactly how many indices you are going to need, and the last and first ones will both be exact.  That code should be easier to follow.